src/searcher/glue.rs

```diff
@@ -0,0 +1,262 @@
+use tree_sitter::{Node, QueryCursor};
+
+use crate::{
+    lines::{self, LineStep},
+    plugin::get_loaded_filter,
+    query_context::QueryContext,
+    searcher::{core::Core, Config, Range, Searcher},
+    sink::Sink,
+    treesitter::{get_parser, node_to_match},
+};
+
+#[derive(Debug, Default)]
+struct AccumulatedExactMatches {
+    matches_with_offsets_relative_to_reference_beginning_of_line_offset: Vec<Range>,
+    reference_beginning_of_line_offset: Option<usize>,
+}
+
+impl AccumulatedExactMatches {
+    pub fn clear(&mut self) {
+        self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+            .clear();
+        self.reference_beginning_of_line_offset = None;
+    }
+
+    pub fn push(
+        &mut self,
+        match_with_absolute_offsets: Range,
+        current_beginning_of_line_offset: usize,
+    ) {
+        if self.reference_beginning_of_line_offset.is_none() {
+            self.reference_beginning_of_line_offset = Some(current_beginning_of_line_offset);
+        }
+        self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+            .push(Range::new(
+                match_with_absolute_offsets.start()
+                    - self.reference_beginning_of_line_offset.unwrap(),
+                match_with_absolute_offsets.end()
+                    - self.reference_beginning_of_line_offset.unwrap(),
+            ));
+    }
+}
+
+impl AsRef<[Range]> for AccumulatedExactMatches {
+    fn as_ref(&self) -> &[Range] {
+        &self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+    }
+}
+
+#[derive(Debug)]
+pub struct MultiLine<'s, S> {
+    config: &'s Config,
+    core: Core<'s, S>,
+    slice: &'s [u8],
+    last_match: Option<Range>,
+    accumulated_exact_matches: AccumulatedExactMatches,
+}
+
+impl<'s, S: Sink> MultiLine<'s, S> {
+    pub fn new(
+        searcher: &'s Searcher,
+        query_context: QueryContext,
+        slice: &'s [u8],
+        write_to: S,
+    ) -> MultiLine<'s, S> {
+        MultiLine {
+            config: &searcher.config,
+            core: Core::new(searcher, query_context, write_to),
+            slice,
+            last_match: None,
+            accumulated_exact_matches: Default::default(),
+        }
+    }
+
+    pub fn run(mut self) -> Result<(), S::Error> {
+        if self.core.begin()? {
+            let mut keepgoing = true;
+            let mut query_cursor = QueryCursor::new();
```

Comment:
Here in our "main searching loop" is where we initialize the tree-sitter query stuff